app/src/test/java/org/mozilla/fenix/ext/StringTest.kt

```diff
@@ -41,4 +54,188 @@ class StringTest {
+    }
+
+    @Test
+    fun `should get the hostname from url`() {
+        "http://bar.com" shortenedShouldBecome "bar.com"
+    }
+
+    @Test
+    fun `should not strip out www if not first subdomain`() {
+        "http://foo.www.com" shortenedShouldBecome "foo.www.com"
+        "http://www.foo.www.com" shortenedShouldBecome "foo.www.com"
+    }
+
+    @Test
+    fun `should convert to lowercase`() {
+        "HTTP://FOO.COM" shortenedShouldBecome "foo.com"
+    }
+
+    @Test
+    fun `should not include the port`() {
+        "http://foo.com:8888" shortenedShouldBecome "foo.com"
+    }
+
+    @Test
+    fun `should return hostname for localhost`() {
+        "http://localhost:8000/" shortenedShouldBecome "localhost"
+    }
+
+    @Test
+    fun `should return hostname for ip address`() {
+        "http://127.0.0.1/foo" shortenedShouldBecome "127.0.0.1"
+    }
+
+    @Test
+    fun `should return etld for www gov uk (www-only non-etld)`() {
+        "https://www.gov.uk/countersigning" shortenedShouldBecome "gov.uk"
+    }
+
+    @Test
+    fun `should return idn etld for www-only non-etld`() {
+        // See comment links on $PUNYCODE and $IDN for term clarification
+        "https://www.$PUNYCODE/foo" shortenedShouldBecome IDN
+    }
+
+    @Test
+    fun `should return not the protocol for file`() {
+        "file:///foo/bar.txt" shortenedShouldBecome "/foo/bar.txt"
+    }
+
+    @Test
+    @Ignore("This behavior conflicts with https://bugzilla.mozilla.org/show_bug.cgi?id=1554984#c11")
+    fun `should return not the protocol for about`() {
+        "about:newtab" shortenedShouldBecome "newtab"
+    }
+
+    @Test
+    fun `should fall back to full url as a last resort`() {
+        "about:" shortenedShouldBecome "about:"
+    }
+    // END test cases borrowed from desktop
+
+    // BEGIN test cases borrowed from FFTV
+    // (https://searchfox.org/mozilla-mobile/source/firefox-echo-show/app/src/test/java/org/mozilla/focus/utils/TestFormattedDomain.java#228)
+    @Test
+    fun testIsIPv4RealAddress() {
+        assertThat("192.168.1.1".isIpv4()).isTrue()
+        assertThat("8.8.8.8".isIpv4()).isTrue()
+        assertThat("63.245.215.20".isIpv4()).isTrue()
+    }
+
+    @Test
+    fun testIsIPv4WithProtocol() {
+        assertThat("http://8.8.8.8".isIpv4()).isFalse()
+        assertThat("https://8.8.8.8".isIpv4()).isFalse()
+    }
+
+    @Test
+    fun testIsIPv4WithPort() {
+        assertThat("8.8.8.8:400".isIpv4()).isFalse()
+        assertThat("8.8.8.8:1337".isIpv4()).isFalse()
+    }
+
+    @Test
+    fun testIsIPv4WithPath() {
+        assertThat("8.8.8.8/index.html".isIpv4()).isFalse()
+        assertThat("8.8.8.8/".isIpv4()).isFalse()
+    }
+
+    @Test
+    fun testIsIPv4WithIPv6() {
+        assertThat("2001:db8::1 ".isIpv4()).isFalse()
+        assertThat("2001:db8:0:1:1:1:1:1".isIpv4()).isFalse()
+        assertThat("[2001:db8:a0b:12f0::1]".isIpv4()).isFalse()
+    }
+    // END test cases borrowed from FFTV
+
```

Comment:
I know you went down the rabbit hole on this one, but I'm not sure if these tests are nearly evil and malicious enough. Maybe some of these are extra, but the URL encoding one could be a very real attack.

- What about URL encoding?
"https://user%40really.evil.domain%2F@mail.google.com" shortenedShouldBecome "really.evil.domain" (not mail.google.com or a crash, both of which are plenty bad)

- What about DWORD IP addresses?
"https://16843009" shortenedShouldBecome "1.1.1.1"

- What about octal IP addresses?
"https://00001.00001.00001.00001" shortenedShouldBecome "1.1.1.1"

- What about hex IP addresses?
"http://0x01010101" shortenedShouldBecome "1.1.1.1"
"http://0x100200401" shortenedShouldBecome "1.1.1.1"

Reply:
Appreciate the extra cases!  IMO, I think we can safely push the last three into a follow up for prioritization (I'll add them as ignored cases, if that's alright with you), but that first one seems _extremely_ important.  I'll start working on fixing it now.

Reply:
I agree with your assessment. The last three are a big can of worms and aren't nearly as important.